[WiP] First steps towards verifying stationarity of a Lagrangian

aryamanjeendgar opened the pull request:
Dummy PR for tracking and reviewing the introduction of the `check_stationarity_lagrangian` method in `solver_test_helpers.py`

aryamanjeendgar committed:
mod stat_check + a skeleton test file

aryamanjeendgar committed:
Fix for concave_expr >= 0 ineq cons

aryamanjeendgar committed:
Tests for QP & SOC's working!

rileyjmurray committed:
add standard test power cone programs to test_KKT. Remove awkward "expect" argument from check_stationary_lagrangian. Modify check_stationary_lagrangian so that it tests gradient norms with respect to individual primal variables, and raise an informative failure message saying what variables had problematic gradients.

rileyjmurray committed:
simplify check_stationary_lagrangian

rileyjmurray committed:
change sign convention for equality constraint dual variables, to simplify how we define Lagrangians. We might want to revert this change.

rileyjmurray committed:
restore sign convention for dual variables from Equality

rileyjmurray committed:
code style

rileyjmurray committed:
remove unnecessary variable name change

aryamanjeendgar committed:
Merge pull request #9 from rileyjmurray/stationarity_checks

Lagrangians: Add power cone program tests, make check_stationary_lagrangian have more informative error messages

aryamanjeendgar committed:
Failing `SDP` tests

aryamanjeendgar commented:
@rileyjmurray @SteveDiamond @phschiele, tagging for review!

aryamanjeendgar committed:
Skeleton for dealing with cons passed via flags

aryamanjeendgar commented on cvxpy/tests/solver_test_helpers.py:
```diff
@@ -110,27 +110,116 @@ def check_complementarity(self, places) -> None:
+            explicitly in the list of constraints passed to the problem --- thus making the above
+            constructed lagrangian incorrect. In these cases instead of checking if the dLdX = 0
+            we check if dLdX \in K^{*} (i.e. the dual cone of the implicit constraint)"""
+            for opt_var in self.prob.variables()
```
Question: Should this pass through the problem variables come before or after the computation of `bad_fro_norms` computation?
If we can figure out a way to compute a `norm` $\to 0$ for every type of 'flag' constraint, then it should definitely be the former

aryamanjeendgar replied:
For `diag` this could just be `np.linalg.norm(np.diag(opt_var.toarray()))`.

For `nonneg` it could be `np.linalg.norm(opt_var.toarray() - np.abs(opt_var.to_array()))`
Similarly, for `nonpos` it could be `np.linalg.norm(opt_var.toarray() + np.abs(opt_var.to_array()))`

But I am not sure what a norm definition which goes to zero for when a matrix is `PSD` would look like. Currently the (PSD) cone membership test that I am using is the existence of the Cholesky Decomposition --- not super sure about how that could be converted into a `norm` $\to 0$.

An alternate approach to characterise is ofc via the spectrum. We could maybe introduce a discrete norm whose value is the number of non-positive eigenvalues of the matrix? (hence being exactly zero when we have all eigenvalues non-negative). Would be super easy to implement as well, though I am not sure if it makes sense to do something like that.

aryamanjeendgar replied:
On @phschiele's suggestion, a good candidate for a quantity in this case would just be the difference between the projection of the matrix variable onto the PSD cone and the matrix itself. 

This is obviously, already done within `residual`! We could also use the same quantity in the case of `nonneg` and `nonpos` since they are also self-dual. My suggestion for `diag` could still be used though.

phschiele commented on cvxpy/tests/test_KKT.py:
```diff
@@ -0,0 +1,216 @@
+    def test_lp_1(self):
+        # typical LP
+        sth = STH.lp_1()
+        if 'MOSEK' in cp.installed_solvers():
+            sth.solve(solver='MOSEK')
+        else:
+            sth.solve(solver='ECOS')
```
Do all tests work with Mosek and Ecos? If so, why do we need this switch?

aryamanjeendgar replied:
Oh, I just included this switch in case a user ran these tests locally and didn’t have `MOSEK` installed. They do indeed pass with both choices.

phschiele replied:
But if they pass with Mosek and Ecos, why include Mosek here at all?
Do you want to test mosek and Ecos, in case a user has both? You could parametrize the tests to do so.

aryamanjeendgar replied:
I explicitly included `MOSEK` because I was testing it out on my system XD. I will parametrize the tests in the final version of the PR!

aryamanjeendgar committed:
Small changes

aryamanjeendgar committed:
More small changes

aryamanjeendgar committed:
Working `check_stationarity_lagrangian`!!!

aryamanjeendgar committed:
Removed redundant `bad_fro_norms` pass & comments

aryamanjeendgar committed:
Minor cleanups

rileyjmurray commented on cvxpy/tests/solver_test_helpers.py:
```diff
@@ -110,27 +110,120 @@ def check_complementarity(self, places) -> None:
             self.tester.assertAlmostEqual(comp, 0, places)
+            
+    def check_stationary_lagrangian(self, places) -> None:
+        L = self.prob.objective.expr
+        for con in self.constraints:
+            if isinstance(con, (cp.constraints.Inequality,
+                                cp.constraints.Equality)):
+                dual_var_value = con.dual_value
+                prim_var_expr = con.expr
+                if self.prob.objective.NAME == 'minimize':
+                    L = L + cp.scalar_product(dual_var_value, prim_var_expr)
+                else:
+                    L = L - cp.scalar_product(dual_var_value, prim_var_expr)
+            elif isinstance(con, (cp.constraints.ExpCone,
+                                  cp.constraints.SOC,
+                                  cp.constraints.Zero,
+                                  cp.constraints.NonNeg,
+                                  cp.constraints.PSD,
+                                  cp.constraints.PowCone3D)):
+                if self.prob.objective.NAME == 'minimize':
+                    L = L - cp.scalar_product(con.args, con.dual_value)
+                else:
+                    L = L + cp.scalar_product(con.args, con.dual_value)
+            else:
+                raise NotImplementedError()
```
Instead of checking if the objective is ``minimize`` when processing constraints, just flip the sign on the objective function. I.e., do
```
    objective = self.prob.objective
    if objective.NAME == 'minimize':
        L = objective.expr
    else:
        L = -objective.expr
```

rileyjmurray commented on cvxpy/tests/solver_test_helpers.py:
```diff
@@ -110,27 +110,120 @@ def check_complementarity(self, places) -> None:
+                raise NotImplementedError()
+        g = L.grad
+        bad_norms = []
+
+        """The reason 'bad' norms could arise (outside of an error by CVXPY) would be if a
+        constraint was introduced in a variable via a flag during it's declaration and not
+        explicitly in the list of constraints passed to the problem --- thus making the above
+        constructed lagrangian incorrect. In these cases instead of checking if the dLdX = 0
+        we check if dLdX \in K^{*} (i.e. the dual cone of the implicit constraint)"""
```
This comment isn't quite right. The Lagrangian is defined by _conventions_. Once we've set certain conventions, it's up to us to interpret the Lagrangian correctly.  Our specific conventions are that conic constraints declared with Variable attributes (PSD=True, diag=True, etc...) are not moved into the Lagrangian. The effect of this convention is that we don't check if the gradient of the Lagrangian is equal to zero, we check if ``dLdX \in K^{*}``. This captures the case when ``K = R^n`` since the dual to ``R^n`` is the singleton set that only contains the zero vector.

rileyjmurray commented on cvxpy/tests/solver_test_helpers.py:
```diff
@@ -110,27 +110,120 @@ def check_complementarity(self, places) -> None:
+                    mat = g_bad_mat + g_bad_mat.T
+                    corrected_bad_norm = np.linalg.norm(mat) / np.sqrt(opt_var.size)
+                    if corrected_bad_norm > 10**(-places):
+                        bad_norms.append((corrected_bad_norm, opt_var))
```
Don't call this ``corrected_bad_norm``.

The purpose of the ``bad_`` prefix in ``bad_norms`` was to signify that the gradient had too large of a component in some direction. If a Variable wasn't subject to any conic constraint, then no direction is allowed, and we only check the norm of the gradient. If a Variable is constrained to lay in ``K``, then the only "disallowed" directions are those outside of the dual ``K^{*}``.

(This comment applies to all subsequent cases as well.)

aryamanjeendgar committed:
Merge branch 'master' into stationarity_checks

aryamanjeendgar committed:
Suggested changes

rileyjmurray commented on cvxpy/tests/solver_test_helpers.py:
```diff
@@ -130,35 +130,36 @@ def check_complementarity(self, places) -> None:
+        `K` is the convex cone corresponding to the implicit constraint on `X`
+        """
         for (opt_var, v) in g.items():
             if all(not attr for attr in list(map(lambda x: x[1], opt_var.attributes.items()))):
```
This if-statement is impossible to decipher. Define a helper function that computes this value. The helper function can be an instance method of the Variable class or it could be a method you define here, in ``solver_test_helpers``.

rileyjmurray replied:
Alternatively, you can just have this case be captured by the "else" statement in this chain of if-else statements.

aryamanjeendgar replied:
Understood! I think the latter would be neater (which would look fine even when we eventually add in support for complex variables and have to deal with the `complex,hermitian` flags)

rileyjmurray commented on cvxpy/tests/solver_test_helpers.py:
```diff
@@ -173,45 +174,45 @@ def check_stationary_lagrangian(self, places) -> None:
-                    corrected_bad_norm = np.linalg.norm(g_bad_mat.residual) / np.sqrt(opt_var.size)
-                    if corrected_bad_norm > 10**(-places):
-                        bad_norms.append((corrected_bad_norm, opt_var))
+                    dual_cone_violation = np.linalg.norm(g_bad_mat.residual) / np.sqrt(opt_var.size)
```
You're trying to compute ``Constant.residual`` here when you should be trying to compute ``Constraint.residual``. That should raise an error. If tests pass with the code as-is then that suggests to me that tests aren't hitting this code path.

aryamanjeendgar replied:
Right! I meant to write `tmp_con.residual` there :)

And yes, there are currently no tests for flag variables (except the SDP ones which incorporate `symmetric=True`) --- which is why I was considering adding a new `StandardTestXXX` class within `test_KKT.py` which included tests for testing out this this `if-else` ladder --- should I do so @rileyjmurray?

rileyjmurray replied:
@aryamanjeendgar, yes, please add such tests. A note: for any explicit constraints you use, there's no need to check dual variables beyond their domains. Complementarity, stationarity, primal feasibility, and feasibility w.r.t. dual domains are sufficient for a correctness check.

rileyjmurray commented:
Tests are failing for Python 3.7 builds on all operating systems. The failures seem to be unrelated to this PR, and they're probably related to the Python Software Foundation dropping support for Python 3.7.

aryamanjeendgar committed:
Added tests for flags + corresponding changes

aryamanjeendgar committed:
Linter fix

rileyjmurray commented:
@aryamanjeendgar you have failing tests on Python 3.8: https://github.com/aryamanjeendgar/cvxpy/actions/runs/5531172776/jobs/10091522330?pr=10#step:6:1668.

rileyjmurray commented on cvxpy/expressions/expression.py:
```diff
@@ -731,6 +731,8 @@ def __lt__(self, other: "Expression"):
 
     @_cast_other
     def __ge__(self, other: "Expression"):
+        """NonPos : Creates an inequality constraint.
+        """
```
Remove this.

rileyjmurray commented on cvxpy/tests/test_KKT.py:
```diff
@@ -0,0 +1,301 @@
+        """
+        X = cp.Variable(shape=(3,3), NSD=True)
+        obj = cp.Maximize(cp.lambda_min(X))
+        cons = [X == X.T, X[0][1] == 123]
```
You don't need ``X == X.T`` if you declared ``NSD=True``. Using ``NSD=True`` or ``PSD=True`` as an attribute enforces a constraint that the variable is symmetric. (This is different from ``X << 0`` or ``X >> 0``, which does not enforce a symmetry constraint automatically.)

rileyjmurray replied:
Don't index into matrix variables by ``X[i][j]``; use ``X[i, j]`` instead. Please fix this here and elsewhere.

aryamanjeendgar replied:
Oh! So, `NSD/PSD=True` explicitly enforces `X == X.T` as part of the set of constraints?

rileyjmurray replied:
Well, I would say it's _implicit_ rather than _explicit_, but yes, it enforces a constraint ``X == X.T``.

aryamanjeendgar replied:
Yeah, the tests still pass even after removing `X == X.T` --- but if the constraint is imposed implicitly (i.e. by simply flipping the `symmetric` flag), meaning that there are now two attributes that are active on the variable (`PSD` & `symmetric`), then the validity of the tests is.... a bit off, right? (since only one of the `if-else` is hit, in this case, the `PSD` one, while for correctness we should be checking for both)

rileyjmurray replied:
You should find that if ``X.is_psd()`` then you also have ``X.is_symmetric()``.
```
>>> import cvxpy as cp
>>> X = cp.Variable(shape=(3,3), PSD=True)
>>> X.is_psd()
True
>>> X.is_symmetric()
True
```

aryamanjeendgar replied:
Yep, that's what I am getting at. Since now `X` has both the attributes `PSD/NSD` and `symmetric` set to True --- however, we only iterate through the variables once and make the checks assuming that every variable would at most have one attribute set to `True` (which is ~mostly true since CVXPY only allows users to set a single parameter manually).

My question is this: We are checking if the gradient corresponding to this variable lies in the dual cone of the imposed implicit constraint --- in this particular example: imposing `symmetric=True` when `PSD/NSD=True` is fair for the variable, but is it also necessary that for some vector $v\in$ `PSD` (since the `PSD/NSD` cones are self-dual) $\implies v\in$ `Skew(M)` --- if not, then we should be checking for BOTH, correct?

rileyjmurray replied:
You're getting at a subtle issue here.

Let $K$ be a convex cone in an inner product space $E$. The dual to $K$ is $K^* = \\{ y : \langle x, y \rangle \geq 0 \forall x \in K\\}$. Note that an inner product appears in this definition

For a pair of convex cones $K_1, K_2$ both contained in an inner product space $E$, if we have $K_1 \subset K_2$ then we have $K_2^* \subset K_1^*$.

It is essential to ask: what inner product space am I using to define the dual cone?
* When $K$ is the PSD cone, we canonically take the dual where $E$ is the space of symmetric matrices.
* When $K$ is the symmetric cone, it makes little sense to use $E$ as the space of symmetric matrices since this would have $K^* = \\{0\\}$. Instead, when $K$ is the symmetric cone, we use $E$ as the space of all square matrices.
* Something to check for yourself: if $K_n$ is the PSD cone for $n \times n$ matrices and you take its dual where $E$ is the space of all $n \times n$ matrices, then you get $K_n^* = \\mathrm{Skew}(n) + K_n$.

rileyjmurray commented on cvxpy/tests/test_KKT.py:
```diff
@@ -0,0 +1,301 @@
+        sth.verify_primal_values(places)
+        sth.check_complementarity(places)
+        sth.verify_dual_values(places)
+        sth.check_stationary_lagrangian(places=4)
```
Is there a reason why you hard-code ``places=4`` here?

aryamanjeendgar replied:
Yeah, some of the other test could only verify correctness till the 2nd decimal place --- but I wanted to check if the stationarity function could give higher accuracy. I'll remove this

rileyjmurray replied:
I actually think testing stationarity to higher accuracy than things like ``verify_primal`` and ``verify_dual`` makes sense. I'd go so far as to say that the only things you need are:
1. ``check_primal_feasibility``
2. ``check_dual_domains``
3. ``check_complementarity``
4. ``check_stationary_lagrangian``
Can you reliably get tests to pass with ``places=4`` for all of these (provided you use the right settings for the solver)?

rileyjmurray commented on cvxpy/tests/test_KKT.py:
```diff
@@ -0,0 +1,301 @@
+
+    @staticmethod
+    def tf_2() -> STH.SolverTestHelper:
+        """
+        Tests PSD flag
+        """
```
Please say how you got the claimed optimal values.

aryamanjeendgar replied:
By solving the problem in CVXPY?

rileyjmurray replied:
With what solver? What settings?

aryamanjeendgar replied:
I'll mention the solver I used, but I didn't really pass in any special solver options (as are defined [here](https://www.cvxpy.org/tutorial/advanced/index.html#setting-solver-options)).

rileyjmurray replied:
Please push changes so I can see changes for this point.

rileyjmurray reviewed (changes requested):
I left several individual comments. Please address and then mark them as resolved.

aryamanjeendgar committed:
More suggested changes

rileyjmurray commented on cvxpy/tests/test_KKT.py:
```diff
@@ -183,13 +183,13 @@ class StandardTestFlags(BaseTest):
         X = cp.Variable(shape=(3,3), NSD=True)
         obj = cp.Maximize(cp.lambda_min(X))
-        cons = [X == X.T, X[0][1] == 123]
+        cons = [X[0][1] == 123]
```
Use ``X[i, j]``, not ``X[i][j]``.

aryamanjeendgar replied:
Oop, mixed this. Made the change!

rileyjmurray commented on cvxpy/tests/test_KKT.py:
```diff
@@ -183,13 +183,13 @@ class StandardTestFlags(BaseTest):
     def tf_1() -> STH.SolverTestHelper:
         """
         Tests NSD flag
+        Reference values via SCS
```
The default termination criteria for SCS are very loose, and they can lead to solutions of mediocre accuracy.

If you use SCS, you should change its termination criteria to be more strict. You should also say what version you used.

aryamanjeendgar replied:
Would re-running these tests with `MOSEK` with the default solver settings be fair?

rileyjmurray replied:
Sure; just say what version of MOSEK you used.

aryamanjeendgar replied:
Re-ran all of these tests using `MOSEK`, and updated the docstring with the version I used. Two of those tests were already run using `MOSEK` XD, I just failed to document that.

aryamanjeendgar committed:
More changes

aryamanjeendgar committed:
Suggested changes

rileyjmurray commented on cvxpy/tests/test_KKT.py:
```diff
@@ -5,175 +5,194 @@
 
 
-class StandardTestFlags(BaseTest):
+class TestKKT_Flags(BaseTest):
```
With these tests it's possible to use much more informative names. I suggest
* ``tf_1`` -> ``nsd_flag``
* ``tf_2`` -> ``psd_flag``
* ``tf_3`` -> ``nonneg_flag``
* ``tf_4`` -> ``nonpos_flag``

(It's fine for the other tests to have names like ``pcp_1``, ``pcp_2``, ``sdp_2``, etc.., because they're actually meant to test a lot of functionality at once.)

Separately, please make sure you have tests that include ``symmetric=True`` and ``diag=True``.

aryamanjeendgar replied:
Understood! Will change those naming conventions.

Also, isn't `symmetric=True` already checked for in the SDP tests? As far as `diag=True` is concerned, we do not support it currently because of how it's not possible to take derivatives w.r.t `diag` variables since they are represented internally as CSC matrices.

rileyjmurray replied:
There are two ways to work with PSD matrices in CVXPY. The first is ``X = cp.Variable(shape=(n,n), PSD=True)``. The second is to declare  ``X = cp.Variable(shape=(n,n), symmetric=True)`` and then explicitly constrain ``X >> 0``. The PSD tests probably only cover one of these. However, they are different for purposes of KKT checks.

It's also possible that a user wants a matrix to be symmetric, but they don't care if it's PSD or NSD. So there should be a test that features ``X = cp.Variable(shape=(n,n), symmetric=True)`` without any semidefinite constraints.

aryamanjeendgar replied:
Understood! I've added a new test (check the `tf_3` in `test_KKT.py`) which has a variable with _just_ the `symmetric=True` flag

rileyjmurray replied:
Thanks. I still need you to address the unclear naming conventions in ``tf_1``, ``tf_2``, .... If someone runs ``pytest cvxpy/tests`` and some of these tests fail, then they should immediately know which parts of CVXPY's functionality is misbehaving (e.g., something related to PSD variables, or symmetric variables, or nonnegative variables).

aryamanjeendgar replied:
Done!

rileyjmurray reviewed (changes requested):
Things are looking much better. I did leave some comments though that I'd like you to address.

aryamanjeendgar committed:
`symmetrc=True` test case

aryamanjeendgar committed:
Informative flag-test names

rileyjmurray committed:
Update test_KKT.py

Naming conventions and tolerances. (These changes are untested and might break in dumb ways. Apologies in advance if that happens.)

rileyjmurray committed:
Update test_KKT.py

Change default places to 4 (from 2) in KKT checks for power cone programs. Made SCS termination criteria more stringent.

rileyjmurray commented:
@aryamanjeendgar you should close this PR, and update your master branch to be in sync with CVXPY's official master branch.
